### PR TITLE
[thunderbird] update images to use new source layout.

### PR DESCRIPTION
**Pull request by tomprince**
This is *completely* untested.

### CODE REVIEW

**Comment by nt1m** [comment]
@tomprince I think it's `./mach` instead of just `mach`.

**Comment by jankeromnes** [comment]
@tomprince That's awesome, thanks a lot for updating the Thunderbird image! :tada: 

> This is *completely* untested.

No worries about that, that's what our CI is for. :smile: Speaking of which, it seems to have choked on Rust install (again), before even reaching the Thunderbird image build:

```
Step 35/52 : RUN rustup install nightly  && rustup completions bash | sudo tee /etc/bash_completion.d/rustup.bash-completion > /dev/null
 ---> Running in 0620f189d470
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: latest update on 2018-06-22, rust version 1.28.0-nightly (662c70a59 2018-06-21)
info: downloading component 'rustc'
error: failed to parse url: 1.28.0-nightly (662c70a59 2018-06-21)/2018-06-22/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
info: caused by: relative URL without a base
The command '/bin/sh -c rustup install nightly  && rustup completions bash | sudo tee /etc/bash_completion.d/rustup.bash-completion > /dev/null' returned a non-zero code: 1
Exited with code 1
```

I'll attempt to re-trigger it, in case this is a temporary Rust hickup (yet another one :stuck_out_tongue:).

File: thunderbird/thunderbird.dockerfile [hunk 1]
```diff
@@ -22,13 +22,14 @@ RUN sudo chown user:user /home/user/thunderbird/.mozconfig
 
 # Set up Mercurial so mach doesn't complain.
 RUN mkdir -p /home/user/.mozbuild \
- && ./mozilla/mach mercurial-setup -u
+ && ln -nsf /home/user/thunderbird/mach /home/user/bin/mach \
```
Comment (jankeromnes) [inline thunderbird/thunderbird.dockerfile]: Hm, I'm not sure `/home/user/bin` is actually in the `PATH`. But a working CI will tell us that. If it's not, we can always add it, or do something like `sudo ln [...] /usr/local/bin/mach`.

**Comment by jankeromnes** [comment]
@tomprince I was finally able to trigger a CI run of your Dockerfile.

Unfortunately, the Thunderbird docker image build failed with this message:

```
Step 4/11 : RUN hg clone --uncompressed https://hg.mozilla.org/mozilla-central/ thunderbird  && cd thunderbird  && hg clone --uncompressed https://hg.mozilla.org/comm-central/ comm WORKDIR thunderbird
 ---> Running in eae97d714a07
applying clone bundle from https://hg.cdn.mozilla.net/mozilla-central/90be04d99fc7941cb9b7186bf5f95e184a4e989a.packed1.hg
452377 files to transfer, 2.14 GB of data
transferred 2.14 GB in 126.5 seconds (17.3 MB/sec)
finished applying clone bundle
searching for changes
adding changesets
adding manifests
adding file changes
added 25 changesets with 127 changes to 113 files
(warning: large working directory being used without fsmonitor enabled; enable fsmonitor to improve performance; see "hg help -e fsmonitor")
new changesets 22764fd6ae90:085cdfb90903
updating to branch default
245233 files updated, 0 files merged, 0 files removed, 0 files unresolved
hg clone: invalid arguments
hg clone [OPTION]... SOURCE [DEST]

make a copy of an existing repository

options ([+] can be repeated):

 -U --noupdate          the clone will include an empty working directory
                        (only a repository)
 -u --updaterev REV     revision, tag, or branch to check out
 -r --rev REV [+]       include the specified changeset
 -b --branch BRANCH [+] clone only the specified branch
    --pull              use pull protocol to copy metadata
    --stream            clone with minimal data processing
 -e --ssh CMD           specify ssh command to use
    --remotecmd CMD     specify hg command to run on the remote side
    --insecure          do not verify server certificate (ignoring web.cacerts
                        config)

(use 'hg clone -h' to show more help)
The command '/bin/sh -c hg clone --uncompressed https://hg.mozilla.org/mozilla-central/ thunderbird  && cd thunderbird  && hg clone --uncompressed https://hg.mozilla.org/comm-central/ comm WORKDIR thunderbird' returned a non-zero code: 255
```

File: thunderbird/thunderbird.dockerfile [hunk 2]
```diff
@@ -11,9 +11,9 @@ RUN sudo apt-get update \
  && sudo rm -rf /var/lib/apt/lists/*
 
 # Download Thunderbird's source code.
-RUN hg clone --uncompressed https://hg.mozilla.org/comm-central/ thunderbird \
+RUN hg clone --uncompressed https://hg.mozilla.org/mozilla-central/ thunderbird \
  && cd thunderbird \
- && python client.py checkout
+ && hg clone --uncompressed https://hg.mozilla.org/comm-central/ comm \
```
Comment (jankeromnes) [inline thunderbird/thunderbird.dockerfile]: Nit: Please don't end a line with `\` if it doesn't continue on the line below.

(This is what caused the CI build to fail.)

**Comment by jankeromnes** [comment]
Thanks a lot for fixing the earlier bug! The image build now failed with:

```
Step 8/12 : RUN mkdir -p /home/user/.mozbuild  && ln -nsf /home/user/thunderbird/mach /home/user/bin/mach  && mach mercurial-setup -u
 ---> Running in ab1bcd5882aa
ln: failed to create symbolic link '/home/user/bin/mach': No such file or directory
The command '/bin/sh -c mkdir -p /home/user/.mozbuild  && ln -nsf /home/user/thunderbird/mach /home/user/bin/mach  && mach mercurial-setup -u' returned a non-zero code: 1
Exited with code 1
```

I guess that's because there is no `/home/user/bin`. Is this a super-widespread convention? (I've never seen it before.) Or maybe you'd like to make the link in `/usr/local/bin` instead?

**Comment by nt1m** [comment]
@jankeromnes @tomprince The new source layout hasn't landed AFAIK: https://dxr.mozilla.org/comm-central/source/

**Comment by tomprince** [comment]
It has definitely landed, dxr and searchfox just haven't been updated to reflect that.

**Comment by nt1m** [comment]
@jankeromnes @tomprince The new source layout has landed. Any update on this ?

**Comment by jankeromnes** [comment]
@nt1m thanks for watching out for unmerged commits!

If it helps, I'm willing to iterate on this and fix build errors, in parallel with fixing the Firefox builds (#211) and the Servo build (following up on #209).

**Comment by jankeromnes** [comment]
(Rebased on top of the latest `master`.)

**Comment by tomprince** [comment]
I'm not actively working on Thunderbird, or using janitor, so if you want to take this over, that'd be great.

**Comment by jankeromnes** [comment]
Hm, looks like we'll have to finish this soon, because it looks like the old layout no longer builds (or it was a temporary issue that I wasn't aware of):

```
Step 3/12 : RUN hg clone --uncompressed https://hg.mozilla.org/comm-central/ thunderbird  && cd thunderbird  && python client.py checkout
 ---> Running in 18d1c733fd4b
applying clone bundle from https://hg.cdn.mozilla.net/comm-central/b027417cfd9fabc5e60db7e57b7d3047a0b99092.packed1.hg
17049 files to transfer, 108 MB of data
transferred 108 MB in 4.8 seconds (22.4 MB/sec)
finished applying clone bundle
searching for changes
adding changesets
adding manifests
adding file changes
added 7 changesets with 278 changes to 244 files
new changesets 47c06b0b47c3:4769902c580e
updating to branch default
8534 files updated, 0 files merged, 0 files removed, 0 files unresolved
python: can't open file 'client.py': [Errno 2] No such file or directory
The command '/bin/sh -c hg clone --uncompressed https://hg.mozilla.org/comm-central/ thunderbird  && cd thunderbird  && python client.py checkout' returned a non-zero code: 2
Exited with code 2
```

**Comment by jankeromnes** [comment]
Rebased and fixed up in #215.

**Pull request closed without merging**